contracts/Verification.sol

```diff
@@ -1,100 +1,88 @@
-        public
-        returns (bool)
-    {
-        // Without this state update the function is gasless
-        stateVariable = stateVariable + 1;
+/**
+ * @title Verifies various hashed data structures
+ */
+contract Verification {
 
-        bytes32 commitment = "";
-        for(uint i = 0; i < _data.length; i++) {
+    /**
+     * @notice Verify a flat array of data elements
+     *
+     * @param _data the array of data elements to be verified
+     * @param _commitment the commitment hash to check
+     * @return a boolean value representing the success or failure of the operation
+     */
+    function verifyMessageArray(
+        bytes32[] calldata _data,
+        bytes32 _commitment
+    ) public pure returns (bool) {
+        bytes32 commitment = _data[0];
+        for (uint256 i = 1; i < _data.length; i++) {
             commitment = keccak256(abi.encodePacked(commitment, _data[i]));
         }
 
-        return _commitment == _commitment;
-    }
-
-    // Data Structure for Set: Merkle Tree with each leaf as a message
-    // Commitment: Hash of the root of the merkle tree
-    // Process for verifying all messages in the set: Take all messages, store them as leaves, build up the entire merkle tree, repeatedly hashing from all leaves to the root.
-    // Process for verifying just a single message in the set: Take a single message as a leaf, and take in hashes of subtrees separate from that leaf and build up just that leaf's branch of the merkle tree, repeatedly hashing from just that leaf to the root.
-
-    struct Tree {
-        bytes32 root;
-        uint256 numbLeaves;
+        return commitment == _commitment;
     }
 
-    Tree tree;
-    mapping(uint256 => bytes32) public leaves;
-    mapping(bytes32 => bytes32) public leafProof;
+    /**
+     * @notice Verify all elements in a Merkle Tree data structure
+     * @dev Performs an in-place merkle tree root calculation
+     * @dev Note there is currently an assumption here that if the number
+     *      of leaves is odd, the final leaf will be duplicated prior to
+     *      calling this function
+     *
+     * @param _data the array of data elements to be verified
+     * @param _commitment the expected merkle root of the structure
+     * @return a boolean value representing the success or failure of the operation
+     */
+    function verifyMerkleAll(
+        bytes32[] memory _data,
+        bytes32 _commitment
+    ) public pure returns (bool) {
+        uint256 hashLength = _data.length;
+
+        for (uint256 j = 0; hashLength > 1; j = 0) {
+            for (uint256 i = 0; i < hashLength; i = i + 2) {
+                _data[j] = keccak256(abi.encodePacked(_data[i], _data[i + 1]));
+                j = j + 1;
+            }
+            // This effectively halves the list length every time,
+            // but a subtraction op-code costs less
+            hashLength = hashLength - j;
+        }
```

Comment:
Makes sense – could you just add a comment on top about the purpose of each loop? That might make this a bit easier to grasp. Thanks 👍